crypto/evp_extra/evp_extra_test.cc

```diff
@@ -3072,12 +3093,279 @@ TEST(EVPExtraTest, ParamgenCallbacks) {
+      {{768, EVP_sha1()}, {2048, EVP_sha224()}, {512, EVP_sha256()}});
+
+  for (std::pair<int, const EVP_MD*> plgen : test_data) {
+    const int nbits = plgen.first;
+    const EVP_MD* digest = plgen.second;
+    // Construct a EVP_PKEY_CTX
+    bssl::UniquePtr<EVP_PKEY_CTX> ctx(
+        EVP_PKEY_CTX_new_id(EVP_PKEY_DSA, nullptr));
+    ASSERT_TRUE(ctx);
+    // Initialize for paramgen
+    ASSERT_TRUE(EVP_PKEY_paramgen_init(ctx.get()));
+    // Set the prime nbits
+    ASSERT_TRUE(EVP_PKEY_CTX_set_dsa_paramgen_bits(ctx.get(), nbits));
+    // Set the digest
+    ASSERT_TRUE(EVP_PKEY_CTX_set_dsa_paramgen_md(ctx.get(), digest));
+
+    EVP_PKEY *raw_pkey = NULL;
+    // Generate the parameters
+    ASSERT_TRUE(EVP_PKEY_paramgen(ctx.get(), &raw_pkey));
+    bssl::UniquePtr<EVP_PKEY> pkey(raw_pkey);
+    ASSERT_TRUE(pkey);
+
+    const DSA* dsa = EVP_PKEY_get0_DSA(pkey.get());
+    ASSERT_TRUE(dsa);
+    const BIGNUM* p = DSA_get0_p(dsa);
+    ASSERT_TRUE(p);
+    unsigned p_size = BN_num_bits(p);
+    ASSERT_EQ(p_size, (unsigned)nbits);
+
+    const BIGNUM* q = DSA_get0_q(dsa);
+    ASSERT_TRUE(q);
+    unsigned q_size = BN_num_bits(q);
+    ASSERT_EQ(q_size, 8*(unsigned)EVP_MD_size(digest));
+  }
+
+  // Test error conditions
+  // Construct a EVP_PKEY_CTX
+  bssl::UniquePtr<EVP_PKEY_CTX> ctx(EVP_PKEY_CTX_new_id(EVP_PKEY_DSA, nullptr));
+  ASSERT_TRUE(ctx);
+  // Initialize for paramgen
+  ASSERT_TRUE(EVP_PKEY_paramgen_init(ctx.get()));
+  // Set the prime length
+  ASSERT_NE(EVP_PKEY_CTX_set_dsa_paramgen_bits(ctx.get(), 511), 1);
+  // Set the generator
+  ASSERT_NE(EVP_PKEY_CTX_set_dsa_paramgen_md(ctx.get(), EVP_md5()), 1);
+}
+
+TEST(EVPExtraTest, DSASignDigestVerify) {
+  bssl::UniquePtr<EVP_PKEY> params = dsa_paramgen(512, EVP_sha1(), false);
+  ASSERT_TRUE(params);
+
+  bssl::UniquePtr<EVP_PKEY> private_key = dsa_keygen(params, false);
+  ASSERT_TRUE(private_key);
+
+  const char data[] = "Sign Me!";
+  const size_t data_len = strnlen(data, 9);
+  uint8_t digest[32] = {0};
+  std::vector<uint8_t> sig;
+  size_t siglen = 0;
+  ASSERT_TRUE(SHA1((uint8_t*)data, data_len, digest));
+  {
+    bssl::UniquePtr<EVP_PKEY_CTX> ctx(EVP_PKEY_CTX_new(private_key.get(), nullptr));
+    ASSERT_TRUE(ctx);
+    ASSERT_TRUE(EVP_PKEY_sign_init(ctx.get()));
+    ASSERT_EQ(1, EVP_PKEY_sign(ctx.get(), NULL, &siglen, digest, 32));
+    ASSERT_GT(siglen, (size_t)0);
+    sig.resize(siglen);
+    ASSERT_EQ(1, EVP_PKEY_sign(ctx.get(), sig.data(), &siglen, digest, 32));
+  }
+
+  bssl::UniquePtr<EVP_PKEY> public_key = dsa_public_key(private_key);
+  {
+    bssl::UniquePtr<EVP_MD_CTX> md_ctx(EVP_MD_CTX_new());
+    ASSERT_EQ(1, EVP_DigestVerifyInit(md_ctx.get(), nullptr, EVP_sha1(), nullptr, public_key.get()));
+    ASSERT_EQ(1, EVP_DigestVerifyUpdate(md_ctx.get(), data, data_len));
+    ASSERT_EQ(1, EVP_DigestVerifyFinal(md_ctx.get(), sig.data(), sig.size()));
+  }
+
+}
+
+TEST(EVPExtraTest, DSADigestSignFinalVerify) {
```

Comment:
Minor nit: can we swap the `EVP_PKEY_sign_init` operations above with `EVP_DigestSignInit` operations below?

That way we'll be testing the same family of APIs: `EVP_DigestSignInit/Update/Final` with `EVP_DigestVerifyInit/Update/Final` and `EVP_PKEY_sign` with `EVP_PKEY_verify`.

Reply:
No problem, should we document that somewhere here? It's hard to tell just from looking at the test and we'll probably want to remember this intention further down the line.